onnxconverter_common/float16.py

```diff
@@ -332,3 +338,58 @@ def convert_float_to_float16_model_path(model_path, min_positive_val=1e-7, max_f
     if not disable_shape_infer:
         model = onnx.load(model_path)
     return convert_float_to_float16(model, min_positive_val, max_finite_val, keep_io_types, disable_shape_infer)
+
+
+def sort_graph_node(graph_proto):
+    # find the "first" node in Nodes that its input is not any node's output
+    def find_first_node(output2node_dict):
+        for node in org_nodes:
+            #if node["name"] not in list_node:
+            is_first_node = True
+            for input in node.input:
+                if input in output2node_dict:  # not a first node
+                    is_first_node = False
+                    break
+            if is_first_node:
+                return node
+        return None
+
+    # remove the node from output2node_dict using output as key
+    def remove_first_node_from_dict2(first_node):
+        for output in first_node.output:
+            del output2node_dict[output]
+
+    org_nodes = graph_proto.node
+    # create a dict to store output as key and node as value
+    output2node_dict = {}
+    for node in org_nodes:
+        for output in node.output:
+            output2node_dict[output] = node
+
+    # save the final node after sorted
+    sorted_node = []
+    # traverse the Nodes to find the first node
+    while(len(output2node_dict) > 0):
+        first_node = find_first_node(output2node_dict)
+        sorted_node.append(first_node)
+        remove_first_node_from_dict2(first_node)
+        # del node from original nodes list to avoid duplicate traverse
+        org_nodes.remove(first_node)
+
+    for new_node in sorted_node:
+        graph_proto.node.extend([new_node])
+
+
+# the input graph should be mode.graph
+# recursevly sort the topology for each sub-graph
+def sort_topology(graph_proto):
```

Comment:
What happens to nodes hosting a subgraph (If, Loop, Scan) and using an input from the main graph?

Reply:
The original code treat the nodes in subgraph (such as If) as global level nodes, and use main graph to do following processing.

Main-graph is entry point,  this function recursively walkthrough from main graph to each sub-graph, the nodes in these graphs will be re-sort according to input/output relationship, say, if a.input is b.output, the node b must be placed before node a. In original code, new added node such as Cast didn't followed this rule, so the inference will be failed.

Reply:
It will be the sub-graph's input, not the output of any node.